### ISSUE TITLE
Internal error NFCeval.evalBinarySub failed to evaluate ‘0 - 0.0‘ with zeros() in nested parameter records and Mutlibody standard library

### ISSUE THREAD

**Issue by afxp** [OpenModelica/OpenModelica#13276]
**tl; dr;**
use `0.0*ones(3)` instead of `zeros(3)` when you need a zeros vector of type Real.

### Description
This is a weird issue and I already have a solution but still would like to share my findings. It is somewhat related to #7387.

### Package
Consider the following package. I did my best to make it as minimal as possible, none can be dropped as far as I tested it. I tried to get rid of the MultiBody Mechanics, but couldn't reproduce the error.

`package Minimal`
`  record ParamsTop`
`    parameter ParamsMid paramsMid;`
`  end ParamsTop;`
` `
`  record ParamsMid`
`    parameter Real mid[3] = zeros(3);`
`    parameter ParamsBottom paramsBottom;`
`  end ParamsMid;`
` `
`  record ParamsBottom`
`    parameter Real bottom[3] = zeros(3);`
`  end ParamsBottom;`
` `
`  record Params`
`    extends ParamsTop(paramsMid=ParamsMid());`
`  end Params;`
` `
`  model Run`
`    parameter Params params;`
`    inner Modelica.Mechanics.MultiBody.World world`
`    Modelica.Mechanics.MultiBody.Parts.FixedTranslation fixedTranslation(r = params.paramsMid.paramsBottom.bottom);`
`  equation`
`    connect(world.frame_b, fixedTranslation.frame_a);`
`  end Run;`
`end Minimal;`

#### Explanation
Model `Run` contains a FixedTranslation (connected to World) whose vector r is set by parameter `bottom`. This parameter is inside nested parameter records. Please don't ask why this complication.

1. `ParamsTop` contains an instance of `ParamsMid`
2. `ParamsMid` contains an instance of `ParamsBottom`, it also contains `mid` not needed to reproduce the error but interesting for investigation
3. `ParamsBottom` contains the parameter vector `bottom`
4. `Params` extends `ParamsTop` and rewrites the `ParamsMid` instance

### Steps to reproduce
First how to reproduce the error with the code above. Translation error pops out of the box. Then change any one point of the next subheadings and the code above works perfectly fine as expected.

#### Original version, translation error

1. Default settings
2. Run `Run`
3. Translation Error `[C:/OM124/OM64bit/OMCompiler/Compiler/NFFrontEnd/NFCeval.mo: 1023:9-1023:67]: Internal error NFCeval.evalBinarySub failed to evaluate ‘0 - 0.0‘`

#### Old frontend, no error

1. Tools -> Options -> Simulation -> check Enable old frontend for code generation
2. Run `Run`
3. No Error

#### Don't rewrite `paramsMid`, no error

1. In `record Params`
  a. Replace `extends ParamsTop(paramsMid=ParamsMid());` with `extends ParamsTop;`
2. Run `Run`
3. No Error

#### Use `mid` instead of `bottom` in `Run`, no error

1. In `model Run`
  a. Replace `(r = params.paramsMid.paramsBottom.bottom)` with `(r = params.paramsMid.mid)`
2. Run `Run`
3. No Error

#### Replace `zeros(3)` with `{0.0,0.0,0.0}` in `Run`, no error

1. In `record ParamsBottom`
  a. Replace `zeros(3)` with `{0.0,0.0,0.0}`
2. Run `Run`
3. No Error

Note that `{0,0,0}` does not solve the error

#### Replace `zeros(3)` with `0.0*ones(3)` in `Run`, no error

1. In `record ParamsBottom`
  a. Replace `zeros(3)` with `0.0*ones(3)``
2. Run `Run`
3. No Error

_This is the preferred error correction._

### Version and OS
 - OpenModelica Version v1.24.0 (64-bit)
 - Modelica Standard Library - Version 4.0.0
 - OS: Windows 10, 64 bit

**Comment by perost** [comment]
Smaller model:
```Modelica
record ParamsBottom
  parameter Real bottom[3] = zeros(3);
end ParamsBottom;

record ParamsMid
  parameter ParamsBottom paramsBottom;
end ParamsMid;

record Params
  parameter ParamsMid paramsMid = ParamsMid();
end Params;

model FixedTranslation
  parameter Params params;
  parameter Real[3] r_shape = {0, 0, 0};
  parameter Real[3] lengthDirection = params.paramsMid.paramsBottom.bottom - r_shape annotation(Evaluate = true);
end FixedTranslation;
```
It seems the compiler forgets that `bottom` should be a `Real` array somewhere and doesn't type cast the result of `zeros(3)`. If the `ParamsMid()` binding equation for `paramsMid` is removed it works, so it seems to be related to evaluating record constructors. I'll take a closer look next week.
